src/wasm/wasm-validator.cpp

```diff
@@ -2767,12 +2767,18 @@ void FunctionValidator::visitFunction(Function* curr) {
   }
 }
 
-static bool checkSegmentOffset(Expression* curr, Address add, Address max) {
+static bool checkSegmentOffset(Expression* curr, Address add, Address max, const FeatureSet &features) {
+  if (!Properties::isValidInConstantExpression(curr, features)) {
+    return false;
+  }
   if (curr->is<GlobalGet>()) {
     return true;
   }
   auto* c = curr->dynCast<Const>();
   if (!c) {
+    if (features.hasExtendedConst()) {
+      return true;
```

Comment:
Shouldn't this check for Add etc. and not accept anything that is not a const? That could be a TODO though for now.

Reply:
isValidInConstantExpression already take care of check only limited instructions are accepted.

This code already gives up if it finds a `global.get` .. so I think its reasonable to also give up in the face of other non-const things (in theory the global.get could result in a const but we don't try to handle this case here).

I'll add a TODO.

Reply:
Added comment.

Reply:
I also restructured the code a bit to make to more clear what is going on here.